src/dbnode/storage/bootstrap/bootstrapper/fs/source.go

```diff
@@ -590,7 +604,14 @@ func (s *fileSystemSource) loadShardReadersDataIntoShardResult(
 				blockStart,
 				blockEnd,
 			)
-			if err != nil {
+			if errors.Is(err, fs.ErrOutOfRetentionClaim) {
+				// Bail early if the index segment is already out of retention.
+				// This can happen when the edge of requested ranges at time of data bootstrap
+				// is now out of retention.
+				s.log.Debug("skipping out of retention index segment", buildIndexLogFields...)
+				s.metrics.persistedIndexBlocksOutOfRetention.Inc(1)
+				return
```

Comment:
for my own understanding: are there any side-effects of partial state being persisted from any shards that had successfully completed?

Reply:
Ah, you mean bootstrapping the TSDB data but not the index data? 

Hm, I think that the data block would not be queryable (if indexing is enabled). But since it is out of retention... it should get offloaded and deleted during the next `Tick()` after bootstrapped.